geode-protobuf/src/main/java/org/apache/geode/internal/protocol/protobuf/v1/operations/PutAllRequestOperationHandler.java

```diff
@@ -60,23 +58,31 @@
     }
 
     long startTime = messageExecutionContext.getStatistics().startOperation();
-    RegionAPI.PutAllResponse.Builder builder = RegionAPI.PutAllResponse.newBuilder()
-        .addAllFailedKeys(putAllRequest.getEntryList().stream()
-            .map((entry) -> singlePut(serializationService, region, entry)).filter(Objects::nonNull)
-            .collect(Collectors.toList()));
-    messageExecutionContext.getStatistics().endOperation(startTime);
+    RegionAPI.PutAllResponse.Builder builder = RegionAPI.PutAllResponse.newBuilder();
+    try {
+      messageExecutionContext.getCache().setReadSerializedForCurrentThread(true);
+
+      for (BasicTypes.Entry entry : putAllRequest.getEntryList()) {
+        BasicTypes.KeyedError error = singlePut(serializationService, region, entry);
+        if (error != null) {
+          builder.addFailedKeys(error);
+        }
+      }
+
+    } finally {
+      messageExecutionContext.getStatistics().endOperation(startTime);
+      messageExecutionContext.getCache().setReadSerializedForCurrentThread(false);
+    }
     return Success.of(builder.build());
   }
 
   private BasicTypes.KeyedError singlePut(SerializationService serializationService, Region region,
-      BasicTypes.Entry entry) {
+      BasicTypes.Entry entry) throws DecodingException {
     try {
       Object decodedValue = serializationService.decode(entry.getValue());
       Object decodedKey = serializationService.decode(entry.getKey());
 
       region.put(decodedKey, decodedValue);
-    } catch (EncodingException ex) {
-      return buildAndLogKeyedError(entry, INVALID_REQUEST, "Encoding not supported", ex);
```

Comment:
The behavior of this operation if a decode exception occurs midway through is now that we'll return an error after having completed some portion of the operation (essentially leaving the region in an undefined state).  This is exactly the behavior that we objected to in the region's putAll function which kept us from being able to use it.

Reply:
I'll write a test for putAll, too, and fix this.